core/embed/upymod/modtrezorio/modtrezorio-hid.h

```diff
@@ -141,6 +141,34 @@ STATIC mp_obj_t mod_trezorio_HID_write(mp_obj_t self, mp_obj_t msg) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_2(mod_trezorio_HID_write_obj,
                                  mod_trezorio_HID_write);
 
+/// def read(self, buf: bytes, offset: int = 0, limit: int | None = None) -> int
+///     """
+///     Reads message using USB HID (device) or UDP (emulator).
+///     """
+STATIC mp_obj_t mod_trezorio_HID_read(size_t n_args, const mp_obj_t *args) {
+  mp_obj_HID_t *o = MP_OBJ_TO_PTR(args[0]);
+  mp_buffer_info_t buf = {0};
+  mp_get_buffer_raise(args[1], &buf, MP_BUFFER_WRITE);
+
+  int offset = 0;
+  if (n_args >= 2) {
+    offset = mp_obj_get_int(args[2]);
+  }
+
+  int limit;
+  if (n_args >= 3) {
+    limit = mp_obj_get_int(args[3]);
+  } else {
+    limit = buf.len - offset;
+  }
+
+  ssize_t r =
+      usb_hid_read(o->info.iface_num, &((uint8_t *)buf.buf)[offset], limit);
```

Comment:
we also need to check that `offset` and `limit` are valid for the `buf`
(and ideally raise a ValueError they exceed)
dtto webusb

Reply:
solved as part of https://github.com/trezor/trezor-firmware/commit/7d9a53c06946b51d4738f5e0c1b8d971f8807b8c